Fix for #193 Missing status when PlugUpdate

starcraftman opened the pull request:
Seems to fix the problem for me. I do not know how long this was an issue though.

@vheon All good for you?

Edit: Work in progress, don't actually pull until I resolve all this.

vheon commented:
Now the status is shown, but when there are some changes the statuses are different, but I don't really know how was before.

In vim:
`- vim-lion:  1 file changed, 20 insertions(+), 4 deletions(-)`

In neovim:
`- vim-lion: Updating 5fa766e..903d4ccFast-forward plugin/lion.vim | 24 ++++++++++++++++++++---- 1 file changed, 20 insertions(+), 4 deletions(-)
`

starcraftman commented:
@vheon Hmmm, doesn't really help that neither of us know what it did before... I just checked and it appears the culprit was this line  `  852     let job.result ,= s:to_s(v:job_data[2])`, we were concatenating the results rather than replacing. Now shows live update. Though, oddly, it ends with just done instead of connectivity..... done.

junegunn commented:
@starcraftman `job.result` holds the entire output of the process, so it is appended. The reason we do this is to show the full message on error. See `s:format_message`.

@vheon I'd like to know why and when the output of neovim was changed. Maybe I can install some older versions of neovim to find out.

starcraftman commented:
@junegunn Ah I see, very well. I changed my branch to preserve the entire output in job.result & simply added job.last_line to track last line to output under normal conditions. Changed s:reap to not clobber buffer with whole job.result. I think this more closely mirrors behaviour of ruby/py live displaying the last line.

starcraftman commented:
Also wow, travis seems to either be overloaded or just de prioritizing free people. Committed 20 mins ago & not one build has run.

junegunn commented on plug.vim:
```diff
@@ -935,7 +940,8 @@ function! s:log(bullet, name, lines)
       let pos = 4
     endif
-    call append(pos - 1, s:format_message(a:bullet, a:name, a:lines))
+    let filtered = substitute(a:lines, '[[:cntrl:]]', '', 'g')
```
So what are the `[[:cntrl:]]` characters we are filtering here? Is this an issue of Neovim? And if so, do you think we should do this as a safety net even when the issue is resolved?

starcraftman replied:
1) When I dump out the contents of a:lines before filtered what I see is most lines end in `\r\n`, while the first few oddly end in `\n\n`. I'm not sure why we are getting Windows EOL when I'm on Linux. So the [[:cntrl:]] could be replaced with [\r\n] if we want to be exact. 

2) Since I don't know what neovim was doing before, can't comment yet. Did older versions only have 1 newline? Trimmed? Perhaps in a recent commit they changed the EOL for some other feature. Might build some older commits to see.

3) Well as a safety issue we could leave it, when using append you aren't ever supposed to leave any new line control characters in the buffer. Got caught by that on Python a few times.

Also this new EOL policy broke s:lines. Quite a nuisance.

starcraftman commented:
@junegunn I just did the bisect, should have done to start with. Commit that breaks our code follows. Apparently, we've been broken for a month.

```
* d28011e (Mon Feb 23 19:43:33 2015) [Thiago de Arruda] eval: Fix buffering of data in job autocommands
| Job autocommands will no longer buffer data chunks that don't end in newlines
| characters.
| 
| 48    50      src/nvim/eval.c
| 25    8       test/functional/job/job_spec.lua
```

starcraftman commented:
@junegunn Related question, why are we appending job.result instead of just storing a list of strings? In the latter case then we can just join when we want more than last. Would immunize against such problems & ease EOL stripping. I need food, will refactor my branch to that I think later.

junegunn commented:
@starcraftman Because neovim installer was written based on the original vimscript implementation where the whole output is obtained using a single `s:system()` call. As long as we keep both versions to share as much code as possible and not deviate much from each other, I'm not against changing it to a list of strings.

starcraftman commented:
@junegunn Alright, I'll see what's easiest to do and ping you when I've got something on my branch worth a look.

junegunn commented:
@starcraftman Thanks! By the way, you want to fix it before neovim changes its API?

starcraftman commented:
@junegunn Condensed it down into a one line fix. Works for me with latest neovim. If you want to you can pull it now. The more major change will be separate, should open a new issue and paste in tarruda's link.

junegunn commented:
Thanks!
